### PR TITLE
Refactor TC exception handling

### PR DESCRIPTION
**Pull request by rail**
* Do not swallow RequestExceptions. The consumer in reflector wraps it
    `try/except` and tclistener won't ack in this case.
  * delete task from the BBB DB in case we get 409 from TC

### CODE REVIEW

**Comment by rail** [comment]
Summoning @catlee

File: bbb/services.py [hunk 1]
```diff
@@ -266,31 +267,44 @@ def start(self):
...
-                                  t.taskId, t.runId)
-                    self.bbb_db.deleteBuildRequest(t.buildrequestId)
-                elif e.superExc.response.status_code == 403:
-                    log.exception("Cannot modify task %s run %s in TC, removing it",
-                                  t.taskId, t.runId)
-                    self.bbb_db.deleteBuildRequest(t.buildrequestId)
-                else:
-                    log.exception("Failed to reflect task %s run %s", t.taskId,
-                                  t.runId)
+                log.exception("Taskcluster exception")
```
Comment (catlee) [inline bbb/services.py]: will this be super noisy now that we're handling all the cases inside _handle_taskcluster_exceptions?

Comment (rail) [inline bbb/services.py]: This is a good point, let me rework it

File: bbb/services.py [hunk 2]
```diff
@@ -266,31 +267,44 @@ def start(self):
...
+            branch = self.buildbot_db.getBranch(t.buildrequestId).split("/")[-1]
+            try:
+                for id_ in self.buildbot_db.getBuildIds(t.buildrequestId):
+                        self.selfserve.cancelBuild(branch, id_)
+            except RequestException:
+                log.exception("Failed to cancel task %s run %s build request ID %s",
+                              t.taskId, t.runId, t.buildrequestId)
+            else:
+                # delete from the DB only if all cancel requests pass
+                self.bbb_db.deleteBuildRequest(t.buildrequestId)
```
Comment (catlee) [inline bbb/services.py]: I think this is correct...but would this make sense as the last line inside the try: section above?

File: bbb/services.py [hunk 3]
```diff
@@ -266,31 +267,44 @@ def start(self):
...
+
+        elif status_code == 404:
```
Comment (catlee) [inline bbb/services.py]: can you add a comment here on why this might happen?

File: bbb/services.py [hunk 4]
```diff
@@ -266,31 +267,44 @@ def start(self):
...
+            log.warn("Cannot find task %s run %s in TC, removing it", t.taskId,
+                     t.runId)
+            self.bbb_db.deleteBuildRequest(t.buildrequestId)
+        elif status_code == 403:
```
Comment (catlee) [inline bbb/services.py]: when would this occur?

File: bbb/services.py [hunk 5]
```diff
@@ -271,30 +271,34 @@ def _handle_taskcluster_exceptions(self, t, exc):
...
         elif status_code == 403:
+            # Bug 1270785. Claiming a completed task returns 403.
             log.warn("Cannot modify task %s run %s in TC, removing it",
                      t.taskId, t.runId)
             self.bbb_db.deleteBuildRequest(t.buildrequestId)
         else:
-            log.exception("Unhandled TC status code %s for task %s run %s",
-                          status_code, t.taskId, t.runId)
+            log.warn("Unhandled TC status code %s for task %s run %s",
+                     status_code, t.taskId, t.runId)
```
Comment (catlee) [inline bbb/services.py]: can you show the exception info in this case, with exc_info=exc?

Comment (rail) [inline bbb/services.py]: The caller function does that actually.

Comment (catlee) [inline bbb/services.py]: yeah...and I was wondering if having that exception logging plus the special case logging here would be too noisy.

Comment (rail) [inline bbb/services.py]: `log.warn()` adds only a single line to already printed traceback. I still want to see the tracebacks until we localize all edge cases.

**Comment by catlee** [comment]
alright, let's shipit!

**Pull request merged**